ext/rotoscope/rotoscope.c

```diff
@@ -138,36 +140,114 @@ static rs_tracepoint_t extract_full_tracevals(rb_trace_arg_t *trace_arg,
+      if (!strcmp(method_name_str, "extend") ||
+          !strcmp(method_name_str, "extend_object") ||
+          !strcmp(method_name_str, "initialize")) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+static VALUE unsafe_obj_method(VALUE argv) {
+  VALUE *args = (VALUE *)argv;
+  return rb_obj_method(args[0], args[1]);
 }
+static VALUE nil_method(VALUE *args) { return Qnil; }
 
-static void log_raw_trace(FILE *stream, rs_tracepoint_t trace) {
-  fprintf(stream, RS_CSV_FORMAT "\n", RS_CSV_VALUES(trace));
+static bool endof_block_method(rb_trace_arg_t *trace_arg) {
+  VALUE method_id = rb_tracearg_method_id(trace_arg);
+  if (NIL_P(method_id)) return false;
+
+  VALUE self = rb_tracearg_self(trace_arg);
+  VALUE argv[2] = {self, method_id};
+  VALUE method = rb_rescue2(unsafe_obj_method, (VALUE)&argv, nil_method,
+                            (VALUE)&argv, rb_eNameError, (VALUE)0);
+  if (NIL_P(method)) return false;
+
+  VALUE iseq = rb_funcall(cInstructionSeq, id_of, 1, method);
+  if (!RTEST(iseq)) return false;
+
+  VALUE label = rb_funcall(iseq, id_label, 0);
+  char *label_str = StringValueCStr(label);
+  return strncmp(VM_BLOCK_PREFIX, label_str, strlen(VM_BLOCK_PREFIX)) == 0;
 }
 
-unsigned char output_buffer[500];
-static void log_stack_frame(FILE *stream, rs_stack_t *stack,
-                            rs_strmemo_t **call_memo, rs_tracepoint_t trace,
-                            rb_event_flag_t event) {
-  if (event & EVENT_CALL) {
-    rs_stack_frame_t frame = rs_stack_push(stack, trace);
-    sprintf((char *)output_buffer, RS_FLATTENED_CSV_FORMAT "\n",
-            RS_FLATTENED_CSV_VALUES(frame));
+static void log_trace_event_with_caller(char *buffer, Rotoscope *config,
+                                        rs_stack_frame_t frame,
+                                        rb_event_flag_t event) {
+  snprintf(buffer, LOG_BUFFER_SIZE, RS_FLATTENED_CSV_FORMAT,
+           RS_FLATTENED_CSV_VALUES(frame));
+  if (!rs_strmemo_uniq(&config->call_memo, (unsigned char *)buffer)) {
+    *buffer = '\0';
+  }
+}
 
-    if (rs_strmemo_uniq(call_memo, output_buffer)) {
-      fputs((char *)output_buffer, stream);
+static void log_trace_event(char *buffer, Rotoscope *config,
+                            rs_tracepoint_t trace, rb_event_flag_t event) {
+  snprintf(buffer, LOG_BUFFER_SIZE, RS_CSV_FORMAT, RS_CSV_VALUES(trace));
+}
+
+static void log_rotoscope_trace(Rotoscope *config, rs_tracepoint_t trace) {
+  rs_stack_frame_t frame;
+  bool return_matches_stack = false;
+  rb_trace_arg_t *trace_arg = rb_tracearg_from_tracepoint(trace.raw);
+  rb_event_flag_t event = rb_tracearg_event_flag(trace_arg);
+
+  *log_buffer = '\0';
+
+  if (METHOD_CALL_P(event)) {
+    frame = rs_stack_push(&config->stack, trace);
+  } else if (METHOD_RETURN_P(event)) {
+    rs_stack_frame_t *last_frame = rs_stack_peek(&config->stack);
+    return_matches_stack = tracecmp(&trace, &last_frame->tp);
+    if (endof_block_method(trace_arg)) {
```

Comment:
I thought doing this for every return event would be a performance problem

Reply:
Surprisingly not. I did some microbenchmarking of all of the newly introduced codepaths, and most evaluations of that method resolved in 1-4μs (with extraordinarily few occurrences—i.e. 2—where it shot up to 80,000 μs), vs `class2str` which occasionally ended up around 10,000μs.

Reply:
Looks like the problem with `class2str` was that we were using `rb_class_path` without first trying to do `rb_class_path` without first trying `rb_class_path_cached`, since `rb_class_path` does `rb_str_dup` on the result.  I pushed a commit to https://github.com/Shopify/rotoscope/pull/44 that will avoid doing that in the future